module/zfs/vdev_removal.c

```diff
@@ -1995,19 +1995,44 @@ spa_vdev_remove_top_check(vdev_t *vd)
 		if (available < vd->vdev_stat.vs_alloc)
 			return (SET_ERROR(ENOSPC));
 	} else {
-		/* available space in the pool's normal class */
-		uint64_t available = dsl_dir_space_available(
-		    spa->spa_dsl_pool->dp_root_dir, NULL, 0, B_TRUE);
-		if (available <
-		    vd->vdev_stat.vs_dspace + spa_get_slop_space(spa)) {
-			/*
-			 * This is a normal device. There has to be enough free
-			 * space to remove the device and leave double the
-			 * "slop" space (i.e. we must leave at least 3% of the
-			 * pool free, in addition to the normal slop space).
-			 */
+		vdev_stat_t *rs = &spa->spa_root_vdev->vdev_stat;
+		vdev_stat_t *vs = &vd->vdev_stat;
+
+		/*
+		 * This is a normal device. There has to be enough free
+		 * space in the pool's normal class to copy the allocated
+		 * space from the device being removed, this will account
+		 * for both reserved and "slop" space.  Any available
+		 * capacity on the device being removed is deducted from
+		 * the total available pool capacity.
+		 */
+		int64_t available = dsl_dir_space_available(
+		    spa->spa_dsl_pool->dp_root_dir, NULL, 0, B_TRUE) -
+		    (vs->vs_dspace - vs->vs_alloc);
```

Comment:
It looks like the new definition of "available" is "space available to the DMU, excluding free space on the removing device", which might be negative.  It's confusing to me that we're adding (via double subtraction) `vs_alloc` here, and then comparing it with `vs_alloc` on line 2034.  It might be more clear to keep the definition of "available" that we have in the current code (i.e. "currently available space", just dsl_dir_space_available()), and compare it with the size of the device to be removed (vs_dspace, as with the current code).

Reply:
Agreed, that was pretty awkward and it look like I can restore the original code here.